From 9ecb6d37ed0749a2fc6608a0b24e41b93225609b Mon Sep 17 00:00:00 2001
From: Lokesh B V <Lokesh.BV@Arm.com>
Date: Wed, 4 Dec 2024 18:40:09 +0530
Subject: [PATCH] neoverse-rd/rdv3r1: enable io block driver

IO block driver represents the layout in which IO virtualisation block
is laid out on the platform. This driver is needed to configure PCIe
rootport attached to the IO virtualisation block.

Add the configuration files for NoC-S3 and IO Block modules and
enable them for scp_ram firmware

Signed-off-by: Lokesh B V <Lokesh.BV@Arm.com>
Change-Id: I998f7c8a5fe6775d9b854363f9ff6ba9afd1ac2f

Signed-off-by: Jon Mason <jon.mason@arm.com>
Upstream-Status: Pending
---
 .../rdv3r1/scp_ramfw/CMakeLists.txt           |   5 +-
 .../rdv3r1/scp_ramfw/Firmware.cmake           |   5 +-
 .../rdv3r1/scp_ramfw/config_io_block.c        | 167 ++++++++++++++++++
 .../rdv3r1/scp_ramfw/config_noc_s3.c          |  21 +++
 .../scp_ramfw/include/io_macro_layout.h       |  80 +++++++++
 .../rdv3r1/scp_ramfw/include/scp_atw0_mmap.h  |   9 +
 .../rdv3r1/scp_ramfw/include/scp_css_mmap.h   |  14 ++
 .../rdv3r1/scp_ramfw/include/scp_fw_mmap.h    |  23 +++
 8 files changed, 322 insertions(+), 2 deletions(-)
 create mode 100644 product/neoverse-rd/rdv3r1/scp_ramfw/config_io_block.c
 create mode 100644 product/neoverse-rd/rdv3r1/scp_ramfw/config_noc_s3.c
 create mode 100644 product/neoverse-rd/rdv3r1/scp_ramfw/include/io_macro_layout.h

diff --git a/product/neoverse-rd/rdv3r1/scp_ramfw/CMakeLists.txt b/product/neoverse-rd/rdv3r1/scp_ramfw/CMakeLists.txt
index 7d41c6fc74c9..d3adec9dc618 100644
--- a/product/neoverse-rd/rdv3r1/scp_ramfw/CMakeLists.txt
+++ b/product/neoverse-rd/rdv3r1/scp_ramfw/CMakeLists.txt
@@ -42,7 +42,10 @@ target_sources(
             "${CMAKE_CURRENT_SOURCE_DIR}/config_pl011.c"
             "${CMAKE_CURRENT_SOURCE_DIR}/config_scp_platform.c"
             "${CMAKE_CURRENT_SOURCE_DIR}/config_system_pik.c"
-            "${CMAKE_CURRENT_SOURCE_DIR}/config_sid.c")
+            "${CMAKE_CURRENT_SOURCE_DIR}/config_sid.c"
+            "${CMAKE_CURRENT_SOURCE_DIR}/config_noc_s3.c"
+            "${CMAKE_CURRENT_SOURCE_DIR}/config_io_block.c"
+    )
 
 #
 # Some of our firmware includes require CMSIS.
diff --git a/product/neoverse-rd/rdv3r1/scp_ramfw/Firmware.cmake b/product/neoverse-rd/rdv3r1/scp_ramfw/Firmware.cmake
index 6280fb2a564a..0aef2f0d80bc 100644
--- a/product/neoverse-rd/rdv3r1/scp_ramfw/Firmware.cmake
+++ b/product/neoverse-rd/rdv3r1/scp_ramfw/Firmware.cmake
@@ -34,7 +34,8 @@ set(SCP_ENABLE_SCMI_NOTIFICATIONS TRUE)
 
 list(PREPEND SCP_MODULE_PATHS
      "${CMAKE_CURRENT_LIST_DIR}/module/scp_platform"
-     "${CMAKE_CURRENT_LIST_DIR}/module/system_pik")
+     "${CMAKE_CURRENT_LIST_DIR}/module/system_pik"
+     "${CMAKE_CURRENT_LIST_DIR}/../module/io_block")
 
 # The order of the modules in the following list is the order in which the
 # modules are initialized, bound, started during the pre-runtime phase.
@@ -46,6 +47,7 @@ list(APPEND SCP_MODULES "system-pik")
 list(APPEND SCP_MODULES "sid")
 list(APPEND SCP_MODULES "atu")
 list(APPEND SCP_MODULES "atu-mmio")
+list(APPEND SCP_MODULES "noc-s3")
 list(APPEND SCP_MODULES "ppu-v1")
 list(APPEND SCP_MODULES "system-power")
 list(APPEND SCP_MODULES "power-domain")
@@ -63,4 +65,5 @@ list(APPEND SCP_MODULES "scmi")
 list(APPEND SCP_MODULES "sds")
 list(APPEND SCP_MODULES "scmi-power-domain")
 list(APPEND SCP_MODULES "scmi-system-power")
+list(APPEND SCP_MODULES "io-block")
 list(APPEND SCP_MODULES "scp-platform")
diff --git a/product/neoverse-rd/rdv3r1/scp_ramfw/config_io_block.c b/product/neoverse-rd/rdv3r1/scp_ramfw/config_io_block.c
new file mode 100644
index 000000000000..cf136888b836
--- /dev/null
+++ b/product/neoverse-rd/rdv3r1/scp_ramfw/config_io_block.c
@@ -0,0 +1,167 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     Configuration data for module 'io_block'.
+ */
+
+#include "io_macro_layout.h"
+#include "scp_css_mmap.h"
+#include "scp_fw_mmap.h"
+
+#include <mod_atu.h>
+#include <mod_io_block.h>
+#include <mod_noc_s3.h>
+#include <mod_scp_platform.h>
+
+#include <fwk_id.h>
+#include <fwk_log.h>
+#include <fwk_macros.h>
+#include <fwk_module.h>
+#include <fwk_module_idx.h>
+
+#define IO_MACRO_ELEMENT_CONFIG(idx) \
+    { \
+        .name = "IO Macro " #idx, \
+        .data = &((struct mod_io_block_element_config) \
+        { \
+            .region_mem_maps_count = 1, \
+            .region_mem_maps = ((struct interface_io_block_setup_mmap[]) \
+            { \
+/*
+* From NOC-S3 TRM:
+* "Warning: The address region that corresponds to the internal configuration
+* space must be programmed in the address map as one of the non-striped
+* regions.
+* The target for this non-striped region must be the
+* Configuration Network Interface (CFGNI) interface ID.
+* If this requirement is not followed, access to the internal configuration
+* completer interface cannot be achieved after the address map has been
+* programmed.”
+*/ \
+                { \
+                    .io_block_address = IO_NOC_S3_GPV_BLOCK_BASE(idx), \
+                    .source_id = ASNI_HNP0_CMN, \
+                    .region_count = 4, \
+                    .carveout_info = ((struct interface_io_block_carveout_info[]){ \
+                        { \
+                            .base = IO_NOC_S3_GPV_BLOCK_BASE(idx), \
+                            .size = IO_NOC_S3_GPV_BLOCK_SIZE, \
+                            .target_id = CFGNI_ID, \
+                            .region_id = 0, \
+                        }, \
+                        { \
+                            .base = TCU_REG_BASE(idx), \
+                            .size = TCU_REG_SIZE, \
+                            .target_id = PMNI_TCU_APB, \
+                            .region_id = 1, \
+                        }, \
+                        { \
+                            .base = CTRL_REG_BASE(idx), \
+                            .size = CTRL_REG_SIZE, \
+                            .target_id = PMNI_CTRL_REG_APB, \
+                            .region_id = 2, \
+                        }, \
+                        { \
+                            .base = RAS_AGENT_REG_BASE(idx), \
+                            .size = RAS_AGENT_REG_SIZE, \
+                            .target_id = PMNI_RAS_AGENT_APB, \
+                            .region_id = 3, \
+                        }, \
+                    }) \
+                }, \
+                { \
+                    .io_block_address = IO_NOC_S3_GPV_BLOCK_BASE(idx), \
+                    .source_id = ASNI_HNP1_CMN, \
+                    .region_count = 4, \
+                    .carveout_info = ((struct interface_io_block_carveout_info[]){ \
+                        { \
+                            .base = IO_NOC_S3_GPV_BLOCK_BASE(idx), \
+                            .size = IO_NOC_S3_GPV_BLOCK_SIZE, \
+                            .target_id = CFGNI_ID, \
+                            .region_id = 0, \
+                        }, \
+                        { \
+                            .base = TCU_REG_BASE(idx), \
+                            .size = TCU_REG_SIZE, \
+                            .target_id = PMNI_TCU_APB, \
+                            .region_id = 1, \
+                        }, \
+                        { \
+                            .base = CTRL_REG_BASE(idx), \
+                            .size = CTRL_REG_SIZE, \
+                            .target_id = PMNI_CTRL_REG_APB, \
+                            .region_id = 2, \
+                        }, \
+                        { \
+                            .base = RAS_AGENT_REG_BASE(idx), \
+                            .size = RAS_AGENT_REG_SIZE, \
+                            .target_id = PMNI_RAS_AGENT_APB, \
+                            .region_id = 3, \
+                        }, \
+                    }) \
+                } \
+            }) \
+        }) \
+    }
+
+static struct mod_io_block_config io_block_module_config = {
+    .noc_s3_id = FWK_ID_MODULE(FWK_MODULE_IDX_NOC_S3),
+    .noc_s3_api_id =
+        FWK_ID_API(FWK_MODULE_IDX_NOC_S3, MOD_NOC_S3_API_SETUP_PSAM),
+    .atu_id = FWK_ID_ELEMENT(FWK_MODULE_IDX_ATU, 0),
+    .atu_api_id = FWK_ID_API(FWK_MODULE_IDX_ATU, MOD_ATU_API_IDX_ATU),
+    .plat_notification = {
+        .notification_id = FWK_ID_NOTIFICATION_INIT(
+                FWK_MODULE_IDX_SCP_PLATFORM,
+                MOD_SCP_PLATFORM_NOTIFICATION_IDX_SUBSYS_INITIALIZED),
+        .source_id = FWK_ID_MODULE_INIT(FWK_MODULE_IDX_SCP_PLATFORM),
+    },
+    .logical_region_base = SCP_ATW0_AP_IO_BLOCK_NOC_S3_BASE,
+    .logical_region_size = SCP_ATW0_AP_IO_BLOCK_NOC_S3_SIZE
+};
+
+static const struct fwk_element io_block_element_table[] = {
+    IO_MACRO_ELEMENT_CONFIG(0),
+    IO_MACRO_ELEMENT_CONFIG(1),
+    { 0 }
+};
+
+void io_block_add_chip_offset(struct mod_io_block_element_config *element_config)
+{
+    uint8_t block_idx = 0;
+    uint8_t region_idx = 0;
+    struct interface_io_block_carveout_info *carveout_info = NULL;
+    uint64_t chip_offset = platform_get_chip_offset();
+    struct interface_io_block_setup_mmap *block_setup_mmap = element_config->region_mem_maps;
+
+    for (block_idx = 0; block_idx < element_config->region_mem_maps_count; block_idx++) {
+        block_setup_mmap[block_idx].io_block_address += chip_offset ;
+        carveout_info = block_setup_mmap[block_idx].carveout_info;
+        for (region_idx = 0; region_idx < block_setup_mmap[block_idx].region_count; region_idx++) {
+            carveout_info[region_idx].base += chip_offset;
+        }
+    }
+}
+
+static const struct fwk_element *io_block_get_element_table(fwk_id_t module_id)
+{
+    uint8_t chip_id = platform_get_chip_id();
+    uint8_t idx = 0;
+
+    FWK_LOG_INFO("[IO Block] chip_id: %d", chip_id);
+
+    for (idx = 0; io_block_element_table[idx].data != NULL; idx++) {
+        io_block_add_chip_offset((struct mod_io_block_element_config *)io_block_element_table[idx].data);
+    }
+
+    return io_block_element_table;
+}
+
+struct fwk_module_config config_io_block = {
+    .data = &io_block_module_config,
+    .elements = FWK_MODULE_DYNAMIC_ELEMENTS(io_block_get_element_table),
+};
diff --git a/product/neoverse-rd/rdv3r1/scp_ramfw/config_noc_s3.c b/product/neoverse-rd/rdv3r1/scp_ramfw/config_noc_s3.c
new file mode 100644
index 000000000000..00b5524ec358
--- /dev/null
+++ b/product/neoverse-rd/rdv3r1/scp_ramfw/config_noc_s3.c
@@ -0,0 +1,21 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     Configuration data for module 'io_block'.
+ */
+
+#include <mod_noc_s3.h>
+
+#include <fwk_element.h>
+#include <fwk_id.h>
+#include <fwk_module.h>
+
+#include <fmw_cmsis.h>
+
+struct fwk_module_config config_noc_s3 = {
+    .elements = FWK_MODULE_STATIC_ELEMENTS({ { 0 } })
+};
diff --git a/product/neoverse-rd/rdv3r1/scp_ramfw/include/io_macro_layout.h b/product/neoverse-rd/rdv3r1/scp_ramfw/include/io_macro_layout.h
new file mode 100644
index 000000000000..3f4ea3180b32
--- /dev/null
+++ b/product/neoverse-rd/rdv3r1/scp_ramfw/include/io_macro_layout.h
@@ -0,0 +1,80 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2024, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     Base address definitions for the SCP's sub-system and access extending
+ *     into the rest of the CSS.
+ */
+
+#ifndef IO_MACRO_LAYOUT
+#define IO_MACRO_LAYOUT
+
+#include "scp_css_mmap.h"
+
+/* Calculate the base address of IO_NOC_S3 GPV for IO Block using its index. */
+#define IO_NOC_S3_GPV_BLOCK_BASE(idx) \
+    (SCP_IO_BLOCK_NOC_S3_GPV_BASE + (idx * SCP_IO_BLOCK_REG_SIZE))
+/* Size of the IO Block IO_NOC_S3 GPV : 16MB */
+#define IO_NOC_S3_GPV_BLOCK_SIZE (16 * FWK_MIB)
+
+/* Calculate the base address for the TCU in IO block using IO block's index. */
+#define TCU_REG_BASE(idx) \
+    (SCP_IO_BLOCK_BASE + (idx * SCP_IO_BLOCK_REG_SIZE))
+#define TCU_REG_SIZE (0x4000000UL)
+
+/* Calculate the base address of IO control register bank. */
+#define CTRL_REG_BASE(idx) \
+    (SCP_IO_BLOCK_CTRL_REG_BASE + (idx * SCP_IO_BLOCK_REG_SIZE))
+#define CTRL_REG_SIZE (0x10000)
+
+/* Calculate the base address of RAS Agent register bank. */
+#define RAS_AGENT_REG_BASE(idx) \
+    (SCP_IO_BLOCK_RAS_AGENT_REG_BASE + (idx * SCP_IO_BLOCK_REG_SIZE))
+#define RAS_AGENT_REG_SIZE (0x10000)
+
+enum asni_domain_id {
+    /*! CMN HNP0 completer Node ID. */
+    ASNI_HNP0_CMN,
+    /*! CMN HNP1 completer Node ID. */
+    ASNI_HNP1_CMN,
+    /*! Maximum number of completer Node IDs*/
+    ASNI_DOMAIN_MAX_ID
+};
+
+enum amni_domain_id {
+    /*! PCIe x16_0 requester Node ID. */
+    AMNI_PCIEX16_0,
+    /*! PCIe x16_1 requester Node ID. */
+    AMNI_PCIEX16_1,
+    /*! PCIe x1_0 requester Node ID. */
+    AMNI_PCIEX1_0,
+    /*! PCIe x4_0 requester Node ID. */
+    AMNI_PCIEX4_0,
+    /*! PCIe x4_1 requester Node ID. */
+    AMNI_PCIEX4_1,
+    /*! PCIe x8_0 requester Node ID. */
+    AMNI_PCIEX8_0,
+    /*! PMNI Control requester Node ID. */
+    PMNI_CTRL_REG_APB,
+    /*! PCIE x16_0 configurtion requester Node ID. */
+    PMNI_PCIEX16_0_CFG,
+    /*! PCIE x1_0 configurtion requester Node ID. */
+    PMNI_PCIEX1_0_CFG,
+    /*! PCIE x4_0 configurtion requester Node ID. */
+    PNI_PCIEX4_0_CFG,
+    /*! PCIE x4_1 configurtion requester Node ID. */
+    PNI_PCIEX4_1_CFG,
+    /*! PCIE x8_0 configurtion requester Node ID. */
+    PMNI_PCIEX8_0_CFG,
+    /*! TCU APB requester Node ID. */
+    PMNI_TCU_APB,
+    /*! RAS Agent Node ID. */
+    PMNI_RAS_AGENT_APB,
+    /*! Config Space interface ID. */
+    CFGNI_ID = 0xF
+};
+
+#endif /* IO_MACRO_LAYOUT */
diff --git a/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_atw0_mmap.h b/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_atw0_mmap.h
index 2eca80a0b112..6ddfd8baf707 100644
--- a/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_atw0_mmap.h
+++ b/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_atw0_mmap.h
@@ -28,6 +28,10 @@
  *   |                                        |
  *   +----------------------------------------+
  *   |                                        |
+ *   | IO NoC S3 (16M)                        |
+ *   |                                        |
+ *   +----------------------------------------+ SCP_ATW0_AP_IO_NOC_S3_BASE
+ *   |                                        |
  *   | MMIO region mapped by ATU (1MB)        |
  *   |                                        |
  *   +----------------------------------------+ SCP_ATW0_ATU_MMIO_BASE
@@ -109,4 +113,9 @@
     (SCP_ATW0_AP_PERIPHERAL_SRAM_BASE + SCP_ATW0_AP_PERIPHERAL_SRAM_SIZE)
 #define SCP_ATW0_ATU_MMIO_SIZE (1 * FWK_MIB)
 
+/* Logical address to access IO Block NoC-S3 via ATU */
+#define SCP_ATW0_AP_IO_BLOCK_NOC_S3_BASE \
+    SCP_ATW0_ATU_MMIO_BASE + SCP_ATW0_ATU_MMIO_SIZE
+#define SCP_ATW0_AP_IO_BLOCK_NOC_S3_SIZE (16 * FWK_MIB)
+
 #endif /* SCP_ATW0_MMAP_H */
diff --git a/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_css_mmap.h b/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_css_mmap.h
index 753befd7f989..069634d10347 100644
--- a/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_css_mmap.h
+++ b/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_css_mmap.h
@@ -37,10 +37,24 @@
 #define SCP_SYSTEM_PIK_BASE (0x50040000)
 #define SCP_PPU_SYS0_BASE (0x50041000)
 
+#define SID_SYS_NODE_NUMBER_MASK UINT32_C(0xFF)
+#define CHIP_ADDRESS_SPACE (64ULL * FWK_GIB)
+
 /* SCP addresses mapped via ATU into address translation windows */
 #define SCP_ADDRESS_TRANSLATION_WINDOW0_BASE (0x60000000)
 #define SCP_ADDRESS_TRANSLATION_WINDOW1_BASE (0xA0000000)
 
+/* Base address of the IO block register bank. */
+#define SCP_IO_BLOCK_BASE (0x1C0000000UL)
+/* Size of the register bank of IO block. */
+#define SCP_IO_BLOCK_REG_SIZE  (0x8000000UL)
+/* Base address of the NoC S3 GPV Space. */
+#define SCP_IO_BLOCK_NOC_S3_GPV_BASE (SCP_IO_BLOCK_BASE + 0x4000000UL)
+/* Base address of the IO Integration Control Registers. */
+#define SCP_IO_BLOCK_CTRL_REG_BASE (SCP_IO_BLOCK_BASE + 0x5B00000)
+/* Base address of the RAS Agent Registers. */
+#define SCP_IO_BLOCK_RAS_AGENT_REG_BASE (SCP_IO_BLOCK_BASE + 0x5B10000)
+
 // clang-format on
 
 #endif /* SCP_CSS_MMAP_H */
diff --git a/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_fw_mmap.h b/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_fw_mmap.h
index 3f7fdc82f32d..35bf94781ea4 100644
--- a/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_fw_mmap.h
+++ b/product/neoverse-rd/rdv3r1/scp_ramfw/include/scp_fw_mmap.h
@@ -59,4 +59,27 @@
 #define SCP_SCMI_PAYLOAD_S_A2P_BASE (SCP_SDS_SECURE_BASE + SCP_SDS_SECURE_SIZE)
 #define SCP_SCMI_PAYLOAD_SIZE       (128)
 
+struct system_sid_reg {
+            uint8_t   RESERVED0[0x40 - 0];
+    FWK_R   uint32_t  SYSTEM_ID;
+            uint8_t   RESERVED1[0x50 - 0x44];
+    FWK_R   uint32_t  SOC_ID;
+            uint8_t   RESERVED2[0x60 - 0x54];
+    FWK_R   uint32_t  NODE_ID;
+            uint8_t   RESERVED3[0x70 - 0x64];
+    FWK_R   uint32_t  SYSTEM_CFG;
+            uint8_t   RESERVED4[0xFD0 - 0x74];
+};
+
+static inline uint8_t platform_get_chip_id(void)
+{
+    struct system_sid_reg *sid_reg = (struct system_sid_reg *)(SCP_SID_BASE);
+    return (sid_reg->NODE_ID & SID_SYS_NODE_NUMBER_MASK);
+}
+
+static inline uint64_t platform_get_chip_offset()
+{
+    return platform_get_chip_id() * CHIP_ADDRESS_SPACE;
+}
+
 #endif /* SCP_FW_MMAP_H */
